src/v/pandaproxy/schema_registry/storage.h

```diff
@@ -1110,29 +1083,65 @@ struct consume_to_store {
             if (!val) {
                 co_await _store.clear_compatibility(*key.sub);
             } else if (key.sub) {
-                co_await _store.set_compatibility(*key.sub, val->compat);
+                co_await _store.set_compatibility(
+                  seq_marker{
+                    .seq = key.seq,
+                    .node = key.node,
+                    .version{invalid_schema_version}, // Not applicable
+                    .key_type = seq_marker_key_type::config},
+                  *key.sub,
+                  val->compat);
             } else {
                 co_await _store.set_compatibility(val->compat);
             }
         } catch (const exception& e) {
-            vlog(plog.debug, "Ignoring: {}: {}", key, e.what());
+            vlog(plog.debug, "Error replaying: {}: {}", key, e);
         }
     }
 
-    ss::future<> apply(delete_subject_key key, delete_subject_value val) {
+    ss::future<> apply(
+      model::offset offset,
+      delete_subject_key key,
+      std::optional<delete_subject_value> val) {
+        // Out-of-place events happen when two writers collide.  First
+        // writer wins: disregard subsequent events whose seq field
+        // doesn't match their actually offset.
+        if (val.has_value() && offset != key.seq) {
+            vlog(
+              plog.info,
+              "Ignoring out of order schema_key {} (at offset {})",
+              key,
+              offset);
+            co_return;
+        }
+
+        if (!val.has_value()) {
+            // This is a tombstone for a previous soft-deletion.  No action
```

Comment:
soft-deletes aren't tombstoned. Can you clarify?

Reply:
It's a tombstone for the delete_subject record, which we emit to release space in the topic but doesn't have any meaning in the context of replay (the presence of the tombstone means a hard delete already happened, we don't need to do anything with the tombstone for the soft delete).

I've reworded the comment for clarity.